src/zalsa.rs

```diff
@@ -377,11 +385,31 @@ where
         db: &'s dyn Database,
         create_index: impl Fn() -> IngredientIndex,
     ) -> &'s I {
-        let zalsa = db.zalsa();
-        let (nonce, index) = self.cached_data.get_or_init(|| {
+        let cached_data = self.cached_data.load(Ordering::Acquire);
+        if cached_data == Self::UNINITIALIZED {
             let index = create_index();
-            (zalsa.nonce(), index)
+            let nonce = db.zalsa().nonce().into_u32().get() as u64;
+            let packed = (nonce << u32::BITS) | (index.0 as u64);
+            debug_assert_ne!(packed, Self::UNINITIALIZED);
+
+            // Discard the result, whether we won over the cache or not does not matter
+            // we know that something has been cached now
+            _ = self.cached_data.compare_exchange(
+                Self::UNINITIALIZED,
+                packed,
+                Ordering::AcqRel,
```

Comment:
`Release` is sufficient here, there's no need for `Acquire` if the current thread wins the race (though of course both are correct).